src/Parsers/IAST.cpp

```diff
@@ -210,14 +200,166 @@ String IAST::getColumnNameWithoutAlias() const
 }
 
+IAST::FormatSettings IAST::FormatSettings::copyWithAlwaysQuoteIdentifiers() const
+{
+    IAST::FormatSettings copy;
+    copy.always_quote_identifiers = true;
+    return copy;
+}
+
+const char * IAST::FormattingBuffer::hilite_keyword      = "\033[1m";
+const char * IAST::FormattingBuffer::hilite_identifier   = "\033[0;36m";
+const char * IAST::FormattingBuffer::hilite_function     = "\033[0;33m";
+const char * IAST::FormattingBuffer::hilite_operator     = "\033[1;33m";
+const char * IAST::FormattingBuffer::hilite_alias        = "\033[0;32m";
+const char * IAST::FormattingBuffer::hilite_substitution = "\033[1;36m";
+const char * IAST::FormattingBuffer::hilite_metacharacter = "\033[1;35m";
+const char * IAST::FormattingBuffer::hilite_none         = "\033[0m";
+
+IAST::FormattingBuffer IAST::FormattingBuffer::copy() const
+{
+    return *this;
+}
+
+const IAST::FormatSettings & IAST::FormattingBuffer::getSettings() const
+{
+    return settings;
+}
+
+IAST::FormattingBuffer IAST::FormattingBuffer::copyWithNewSettings(const IAST::FormatSettings & settings_) const
+{
+    return IAST::FormattingBuffer(ostr, settings_, state, stacked_state);
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::setIndent(UInt8 indent)
+{
+    stacked_state.indent = indent;
+    return *this;
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::increaseIndent(int extra_indent)
+{
+    stacked_state.indent += extra_indent;
+    return *this;
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::setNeedParens(bool value)
+{
+    stacked_state.need_parens = value;
+    return *this;
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::setExpressionListAlwaysStartsOnNewLine(bool value)
+{
+    stacked_state.expression_list_always_start_on_new_line = value;
+    return *this;
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::setExpressionListPrependWhitespace(bool value)
+{
+    stacked_state.expression_list_prepend_whitespace = value;
+    return *this;
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::setSurroundEachListElementWithParens(bool value)
+{
+    stacked_state.surround_each_list_element_with_parens = value;
+    return *this;
+}
+
+IAST::FormattingBuffer & IAST::FormattingBuffer::setCurrentSelect(const IAST * select)
+{
+    stacked_state.current_select = select;
+    return *this;
+}
+
+IAST::FormattingBuffer::Hiliter::Hiliter(DB::WriteBuffer & ostr_, bool hilite_, const char * hilite_type) : ostr(ostr_), hilite(hilite_)
+{
+    if (hilite)
+        ostr << hilite_type;
+}
+
+IAST::FormattingBuffer::Hiliter::~Hiliter()
+{
+    if (hilite)
+        ostr << IAST::FormattingBuffer::hilite_none;
+}
+
+IAST::FormattingBuffer::Hiliter IAST::FormattingBuffer::createHiliter(const char * hilite_type) const
+{
+    return {ostr, settings.hilite, hilite_type};
+}
+
+void IAST::FormattingBuffer::writePossiblyHilited(std::string_view str, const char * hilite_type) const
+{
+    Hiliter hiliter = createHiliter(hilite_type);
+    ostr << str;
```

Comment:
A possible optimization: instead of writing a closing hiliter in `~Hiliter` `FormattingBuffer` can remember the last opening hiliter and check it with next output. For example, if you have a sequence of `writeKeyword()` it would allow to write only one `hilite_keyword` and only one `hilite_none`.

Reply:
I was thinking about but introducing a state machine would complicate the logic for no gain, as there's no need for the optimisation of hilite. There are so many public methods which will need to flush the buffer.

And if anyone ever goes back to implementing the original task of limiting line length, then it will become even more complicated, as it will probably need a state machine of its own.

Reply:
> And if anyone ever goes back to implementing the original task of limiting line length, then it will become even more complicated, as it will probably need a state machine of its own.

What is the original task?

Reply:
[Limit line length in clickhouse-format command #38317
](https://github.com/ClickHouse/ClickHouse/issues/38317)

Reply:
In regard to https://github.com/ClickHouse/ClickHouse/issues/38317 I still don't see why it's better to format
```
\u001b[1mSHOW \u001b[0m\u001b[1m\u001b[0m\u001b[1mTABLES\u001b[0m
```
instead of
```
\u001b[1mSHOW TABLES\u001b[0m
```

In order to limit the length of a line the characters must be counted without hilite sequences anyway, I suppose. So `FormattingBuffer` needs to memorize the current number of characters in the current line. And I don't think there is a lot of difference here between memorizing one or two variables.

Reply:
I feel like it's a premature optimisation. It doesn't affect performance, it only complicates the code logic.

I agree that squashing hilites together does improve readability for debugging, though. But it's a very abstract scenario.

However, I just realised that implementing squashing might over-complicate the whole PR. Earlier we agreed that I'll encapsulate `FormattingBuffer.ostr` in a separate PR because it affects ~1000 lines.

To summarize, here are pros for squashing hilites:
- easier readability for debugging
- XX% fewer bytes per hilited query (but does anyone care?)

Cons:
- complicates the code (additional state machine)
- complicates the PR (encapsulating `ostr` here)
- I'll need to write more code instead of keeping everything as is :)

The question is, do you feel like the pros outweigh the cons? I'm honestly not very excited to write even more code at this point, and I feel like just merging this PR as is would already bring value.

Reply:
> complicates the PR (encapsulating ostr here)

We can encapsulate ostr in a half-way manner:

```
class FormattingBuffer
{
public:
    ~FormattingBuffer() { setHilite(hilite_none); }

    WriteBuffer & out() { setHilite(hilite_none); return ostr; }

    void writeKeyword(std::string_view str) const
    {
        if (keyword.empty())
            return;
        setHilite(hilite_keyword);
        ostr << str;
    }

private:
    WriteBuffer & ostr;
    ...
};
```

and also we will need to mechanically replace every `.ostr` with `.out()`. It doesn't seem like a difficult task.

> I feel like it's a premature optimisation. It doesn't affect performance, it only complicates the code logic.

I'm not 100% sure about performance, but anyway. We format ASTs very often, it's not like one-time task. So it would be nice to be accurate with that.